components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+// TODO: doc for each public func in protocols.
+protocol FxaAccountManagerProtocol {
+    init(config: FxAConfig, deviceConfig: DeviceConfig, applicationScopes: [String], keychainAccessGroup: String?)
+
+    // Starts the FxA account manager and advances the state machine.
+    // It is required to call this method before doing anything else with the manager.
+    func initialize(completionHandler: @escaping (Result<Void, Error>) -> Void)
+
+    func hasAccount() -> Bool
+    func accountProfile() -> Profile?
+    func refreshProfile()
+    func accountNeedsReauth() -> Bool
+
+    func beginAuthentication(completionHandler: @escaping (Result<URL, Error>) -> Void)
+    func beginPairingAuthentication(pairingUrl: String, completionHandler: @escaping (Result<URL, Error>) -> Void)
+    func finishAuthentication(authData: FxaAuthData, completionHandler: @escaping (Result<Void, Error>) -> Void)
+    func getAccessToken(scope: String, completionHandler: @escaping (Result<AccessTokenInfo, Error>) -> Void)
+
+    func deviceConstellation() -> DeviceConstellationProtocol?
+
+    func logout(completionHandler: @escaping (Result<Void, Error>) -> Void)
+}
+
+public extension Notification.Name {
+    static let accountLoggedOut = Notification.Name("accountLoggedOut")
+    static let accountAuthProblems = Notification.Name("accountAuthProblems")
+    static let accountAuthenticated = Notification.Name("accountAuthenticated")
+    static let accountProfileUpdate = Notification.Name("accountProfileUpdate")
+}
+
+extension Notification.Name {
+    static let accountAuthException = Notification.Name("accountAuthException")
+}
+
+protocol AccountStorage {
+    func read() -> FirefoxAccount?
+    func write(_ json: String)
+    func clear()
+}
+
+// swiftlint:disable type_body_length
+open class FxaAccountManager: FxaAccountManagerProtocol {
+    // `var` for testing.
+    var accountStorage: AccountStorage
+    let config: FxAConfig
+    let deviceConfig: DeviceConfig
+    let applicationScopes: [String]
+
+    var acct: FirefoxAccount?
+    var account: FirefoxAccount? {
+        get { return acct }
+        set {
+            acct = newValue
+            if let acc = acct {
+                constellation = makeDeviceConstellation(account: acc)
+            }
+        }
+    }
+
+    var state: AccountState = AccountState.start
+    var profile: Profile?
+    var constellation: DeviceConstellation?
+    var latestAuthState: String?
```

Comment:
Took me a while to figure out that this is not a "state" in the sense of the state-machine, but rather than OAuth "state" parameter. Naming it something like `latestOAuthStateParam` or similar might make that a bit clearer to future readers.